FEATURE BRANCH: Add `v2beta1` api version: Consolidate schema breaking changes

dprokop opened the pull request:
Includes:
  - https://github.com/grafana/grafana/pull/108121
  - DS consolidation changes https://github.com/grafana/grafana/pull/107985
  - DS consolidation changes https://github.com/grafana/grafana/pull/107984
  - https://github.com/grafana/grafana/pull/108177
  - https://github.com/grafana/grafana/issues/106877
  - https://github.com/grafana/grafana/issues/106878

---

Fixes https://github.com/grafana/grafana/issues/108127



**🚨 Dashboards V2 API/Schema Breaking Changes - PR #108172**
**What:** New `v2beta1` API version consolidating dashboard schema changes  

### 1. API Version Update
- `v2beta1` is now default (was `v2alpha1`) when `dashboardNewLayouts` toggle is enabled. 
- Schema imports changed to `@grafana/schema/dist/esm/schema/dashboard/v2`

### **2. DataQueryKind Structure** *(from [PR #104194](https://github.com/grafana/grafana/pull/104194))*
**Major restructuring for future-proofing query and data source model**

Variables (Query/GroupBy/AdHoc), panel queries, and annotations now use new format:
```typescript
// Before: 
{ 
  kind: "prometheus", 
  spec: { expr: "..." } 
}

// After:  
{ 
  kind: "DataQuery", 
  group: "prometheus", 
  version: "v0",
  datasource?: { name: "uid" },  // optional UID reference
  spec: { expr: "..." } 
}
```

**Key changes:**
- `kind` is now always `"DataQuery"`
- Data source type moved to `group` field
- Added `version` for query schema versioning (_in the future_)
- Enhanced `datasource` field supports both stateful (UID) and stateless queries
- Enables future label-based data source resolution (_in the future_)

### **3. VizConfigKind Refactor**
Panel configs restructured:
```typescript
// Before: 
{ 
  kind: "stat", 
  spec: { pluginVersion, options, fieldConfig } 
}

// After:  
{ 
  kind: "VizConfig", 
  group: "stat", 
  version: "pluginVersion", 
  spec: { options, fieldConfig } 
}
```

**Migration:** This is a **forward-compatible change** - new dashboards use the updated schema, but older systems may not understand the new format. I.e. instance running v2alpha1 will not understand v2beta1.

dprokop committed:
Revert "Revert: Future-proofing query and data source model in Dashboard Sche… (#107985)"

This reverts commit 13a89d4ae3f1f4ccdb98edc9b42f3c35522d2dfa.

dprokop committed:
Revert "Revert "Schema V2: Simplify annotations v1<->v2 conversions" (#107984)"

This reverts commit 2b8c5bea1a6e5a54d761ce3e7bc7f5a1c529f60e.

dprokop committed:
make gen apps

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into schemav2-bring-back-reverts

dprokop committed:
Merge branch 'main' into schemav2-bring-back-reverts

dprokop committed:
e2e update

ivanortegaalba committed:
Use v2alpha2 by default (#108177)

* Use v2alpha2 by default

* Apply only DS changes to alpha2

* Use v2alpha2 by default except to query

* Create a v2 index in @grafana/schema

* Update path and apply lint

* Update tests

ivanortegaalba committed:
Update imports to v2 status

dprokop committed:
Fix failing openapi test

dprokop committed:
Merge remote-tracking branch 'origin/main' into schemav2-bring-back-reverts

dprokop committed:
Merge branch 'main' into schemav2-bring-back-reverts

dprokop committed:
Schemav2 breaking changes: conversion implementation (#108224)

* provision v2alpha1 dashboard

* Run conversions for DS refactor

* Run snapshot testing on conversions

* Normalize output name

* Update snapshots to include all panel and variable cases

* fix lint

* fix lint

* fix test and go lint

* more go lint

---------

Co-authored-by: Ivan Ortega <ivanortegaalba@gmail.com>
Co-authored-by: Haris Rozajac <haris.rozajac12@gmail.com>

harisrozajac commented on e2e-playwright/dashboards/TestV2Dashboard.json:
```diff
@@ -2,41 +2,32 @@
   "apiVersion": "dashboard.grafana.app/v2alpha1",
```
```suggestion
  "apiVersion": "dashboard.grafana.app/v2alpha2",
```

dprokop committed:
Merge remote-tracking branch 'origin/main' into schemav2-bring-back-reverts

dprokop committed:
Schema v2: Introduce group/datasource convention to GroupBy and AdHoc variable (#108237)

* Schema v2: Introduce group/datasource convention to GroupBy and AdHoc variables

* add conversion

* App Installer: Authorizer support (#108419)

* Chore: use `satisfies` and remove a load of `any`s (#108397)

use satisfies and remove a load of anys

* improve logging and fail unified-storage migration with more than 0 errors (#108471)

improve logging and fail unified-storage migration with more than 0 errors

* fix conversion test

* Secrets: Create more granular fixed roles for SecureValues (#108382)

* Provisioning: Fix bug in job progress recording (#108440)

Fix bug in job progress recording

* Provisioning: Fix ImportAllPanelsFromLocalRepository test (#108441)

* Provisioning: Skip flaky test

* Fix flaky provisioning test

* Fix lint

---------

Co-authored-by: Roberto Jimenez Sanchez <roberto.jimenez@grafana.com>

* BulkDeleteProvisionedResource: Move progress bar into a second step (#108417)

* Move progress bar into a second step

---------

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

* [Dashboard Schema Codegen] Move dashboard CUE codegen block back up into kind body (#108476)

[Dashboard Schema Codegen] Move dashboard CUE codegen block back up into kind body to make sure new versions have the same settings.

---------

Co-authored-by: Haris Rozajac <haris.rozajac12@gmail.com>
Co-authored-by: Todd Treece <360020+toddtreece@users.noreply.github.com>
Co-authored-by: Ashley Harrison <ashley.harrison@grafana.com>
Co-authored-by: Will Assis <35489495+gassiss@users.noreply.github.com>
Co-authored-by: Matheus Macabu <macabu@users.noreply.github.com>
Co-authored-by: Roberto Jiménez Sánchez <jszroberto@gmail.com>
Co-authored-by: Roberto Jimenez Sanchez <roberto.jimenez@grafana.com>
Co-authored-by: Yunwen Zheng <yunwen.zheng@grafana.com>
Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>
Co-authored-by: Austin Pond <IfSentient@users.noreply.github.com>
Co-authored-by: Ivan Ortega <ivanortegaalba@gmail.com>

harisrozajac committed:
Dashboard Schema V2: Refactor VizConfigKind to follow DataQueryKind convention (#108148)

* Dashboards API: Register v2alpha2 API

* Prepare conversion functions

* Fix test

* Refactor VizConfigKind to follow DataQueryKind convention

* fix tests

* use new dataquerykind convention alpha 2

* add conversion

* fix tests

* fix tests

* fix another test

* Fix merge

---------

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

harisrozajac committed:
fix k8s codegen

ivanortegaalba commented on apps/alerting/notifications/pkg/apis/alerting_manifest.go:
We should skip this from this PR

ivanortegaalba commented on apps/alerting/notifications/Makefile:
We should skip this from this PR, and everything under `apps/alerting/`

ivanortegaalba commented on apps/folder/pkg/apis/folder_manifest.go:
We should skip this from this PR

ivanortegaalba commented on apps/secret/pkg/apis/secret_manifest.go:
We should skip this from this PR

ivanortegaalba commented on go.work.sum:
Do we need this?

ivanortegaalba commented on pkg/registry/apis/dashboard/legacy/sql_dashboards.go:
```diff
@@ -175,7 +175,8 @@ func (r *rowsWrapper) Next() bool {
 		r.row, err = r.a.scanRow(r.rows, r.history)
 		if err != nil {
-			if len(r.rejected) > 1000 || r.row == nil {
+			r.a.log.Error("error scanning dashboard", "error", err)
+			if len(r.rejected) > 0 || r.row == nil {
```
```suggestion
			if len(r.rejected) > 0 || r.row == nil {
```

ivanortegaalba replied:
Missing console log

ivanortegaalba commented on pkg/registry/apis/provisioning/jobs/driver.go:
Where do these changes come from?

ivanortegaalba commented on pkg/registry/apis/secret/accesscontrol.go:
Where do these changes come from?

ivanortegaalba commented on pkg/services/apiserver/appinstaller/installer.go:
I don't understand all changes under `services/apiserver`

dprokop replied:
me neither. I think we should reset this to `main`

ivanortegaalba commented on public/app/plugins/datasource/tempo/traceql/traceql.test.ts:
Everything withing plugins/datashould shouldn't be included

ivanortegaalba commented on public/app/features/provisioning/Shared/ProgressBar.tsx:
Why all of this is here?

ivanortegaalba commented on public/locales/en-US/grafana.json:
This probably shouldn't be here

harisrozajac committed:
Merge branch 'main' into schemav2-bring-back-reverts

ivanortegaalba reviewed (commented):
Let's remove unnecessary changes from other apps/* and code-wise looks good to me!

harisrozajac committed:
Update e2e-playwright/dashboards/TestV2Dashboard.json

harisrozajac committed:
Update e2e/dashboards/TestV2Dashboard.json

harisrozajac committed:
revert app generation for non-related apps

harisrozajac commented on apps/alerting/notifications/pkg/apis/alerting/v0alpha1/receiver_schema_gen.go:
```diff
@@ -13,7 +13,7 @@ import (
 var (
 	schemaReceiver = resource.NewSimpleSchema("notifications.alerting.grafana.app", "v0alpha1", &Receiver{}, &ReceiverList{}, resource.WithKind("Receiver"),
-		resource.WithPlural("receivers"), resource.WithScope(resource.NamespacedScope), resource.WithSelectableFields([]resource.SelectableField{resource.SelectableField{
+		resource.WithPlural("receivers"), resource.WithScope(resource.NamespacedScope), resource.WithSelectableFields([]resource.SelectableField{{
```
@dprokop something weird happens with these files, because when I run `make generate` for alerting locally, I get the output that should revert this change, but when I commit, this doesn't update. Same with folder and secrets apps

harisrozajac replied:
nvm, it worked when I commented out this part in lefthook.yml:
```
backend-format:
    glob: '*pkg/**/*.go'
    run: gofmt -w -s {staged_files}
    stage_fixed: true
```

harisrozajac committed:
try again

harisrozajac committed:
another try

harisrozajac committed:
also revert folder and secret app generation

harisrozajac committed:
Merge branch 'main' into schemav2-bring-back-reverts

ivanortegaalba commented on apps/dashboard/pkg/migration/conversion/testdata/output/v2alpha1.ds-data-query.v2alpha2.json:
```diff
@@ -0,0 +1,988 @@
+      "fiscalYearStartMonth": 0
+    },
+    "title": "Test: V2alpha1 dashboard with annotations",
+    "variables": [
+      null,
+      null,
+      null,
+      null
+    ]
```
There are no conversions for variables!

dprokop committed:
v2alpha1 provisioned dashboard

dprokop committed:
Fix kind

dprokop committed:
Fix conversion snapshots

harisrozajac committed:
Merge branch 'main' into schemav2-bring-back-reverts

ivanortegaalba reviewed (approved):
Double checked, it looks great! I'd love the eyes of @stephaniehingtgen from the backend point of view, but looks great so far 💯

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into schemav2-bring-back-reverts

# Conflicts:
#	go.work.sum

ivanortegaalba committed:
Update API discovery registry

ivanortegaalba committed:
Rename to v2beta1

ivanortegaalba committed:
Rename migrations

harisrozajac commented:
@ivanortegaalba should we wait for app platform's input?

stephaniehingtgen reviewed (approved):
LGTM overall - one small change & one question

harisrozajac committed:
Update apps/dashboard/pkg/apis/dashboard/v2beta1/doc.go

Co-authored-by: Stephanie Hingtgen <stephanie.hingtgen@grafana.com>

ivanortegaalba commented on apps/dashboard/pkg/migration/conversion/v2alpha1_to_v2beta1.go:
```diff
@@ -0,0 +1,939 @@
+	if in == nil {
+		return nil
+	}
+	// For brevity, assume the structure is identical and just return nil
+	// In a real implementation, this would need proper conversion
```
😮 I missed it! I'm going to add a test to ensure everything is converted

ivanortegaalba committed:
Ensure conditional rendering and other non changed properties

harisrozajac reviewed (approved):
🚀
